### DISCUSSION TITLE
randomize blocks between participants

### DISCUSSION THREAD

**Discussion by lily2brain** [jspsych/jsPsych#1663] [Q&A]
Hi guys
I am having difficulties to randomize 2 blocks (blocktrial_1_free and blocktrial_2_fixed), the two have different configurations and the information within each block is already randomized, but I need the presentation of these blocks to be random for the participant (ex: participant 1 - blocktrial_2_fixed and blocktrial_1_free / participant 2 - blocktrial_1_free and blocktrial_2_fixed) ... part of the script is below for better understanding:

```js
       // create blockstimuli_1_free
	var blockstimuli_1_free = [
      	  {stimulus: 'image1.jpg', prompt: "<p>Qual?</p>", data: {correct_response: '1'}},
	      {stimulus: 'image2.jpg', prompt:"<p>Qual?</p>", data: {correct_response: '0'}}, 
	];
	
	// create blocktrial_1_free
	var blocktrial_1_free = {
		type: "image-button-response", 
		stimulus: jsPsych.timelineVariable('stimulus'),
		prompt: jsPsych.timelineVariable('prompt'),
		stimulus_width: 950,
		choices:["<p><strong>E</strong></p>", "<p><strong>D</strong></p>"],
		margin_horizontal: '250px',
		
		data: jsPsych.timelineVariable('data'),
		on_finish: function(data){
			data.correct = data.response == data.correct_response;
		}
	};
	
	// create blocktrial_procedure_1_free
	var blocktrial_procedure_1_free = {
		timeline: [fixation, blocktrial_1_free],
		timeline_variables: blockstimuli_1_free, 
	    randomize_order: true
	};
	timeline.push(blocktrial_procedure_1_free);
	
	// create blockstimuli_2_fixed
	var blockstimuli_2_fixed = [
      {stimulus: 'image3.jpg', prompt:"<p>Qual?</p>", data: {correct_response: '1'}},
	  {stimulus: 'image4.jpg', prompt: "<p>Qual?</p>", data: {correct_response: '0'}} 
	];
	
	// create blocktrial_2_fixed
	var blocktrial_2_fixed = {
		type: "image-button-response",
		stimulus: jsPsych.timelineVariable('stimulus'),
		prompt: jsPsych.timelineVariable('prompt'),
		stimulus_width: 950,
		stimulus_duration: 2000,
		choices:["<p><strong>E</strong></p>", "<p><strong>D</strong></p>"],
		margin_horizontal: '250px',
		
		data: jsPsych.timelineVariable('data'),
		on_finish: function(data){
			data.correct = data.response == data.correct_response;
		}
	};
	
	// create blocktrial_procedure_2_fixed
	var blocktrial_procedure_2_fixed = {
		timeline: [fixation, blocktrial_2_fixed],
		timeline_variables: blockstimuli_2_fixed, 
	    randomize_order: true
	};
	timeline.push(blocktrial_procedure_2_fixed);
```

**Comment by becky-gilbert** [comment]
One way to do this is to use [jsPsych.randomization.shuffle](https://www.jspsych.org/core_library/jspsych-randomization/#jspsychrandomizationshuffle) to randomize the order of the two procedures, and then add the procedures to the timeline in this randomized order:
```js
// this will give you an array with the two procedures in random order
var random_block_order = jsPsych.randomization.shuffle([blocktrial_procedure_1_free, blocktrial_procedure_2_fixed]);
// add the first procedure from the randomized array to your timeline
timeline.push(random_block_order[0]);
// add the second procedure from the randomized array to your timeline
timeline.push(random_block_order[1]);
```
I hope this helps. Let us know if this works and if you have any other questions.

**Reply by lily2brain** [reply]
Hello, I tried it the way you told me, but it didn't work, could you help me with another alternative?

**Reply by becky-gilbert** [reply]
Hi @lily2brain, can you explain what you mean when you say it doesn't work? Does the experiment not run or does the randomizing not work? When you open the browser's developer tools, do you see any errors in the console?

**Reply by lily2brain** [reply]
Hi, the randomization that didn't work, and on the console there was no error. Together with my teacher, we managed this way:


//sample experiment version
var version = jsPsych.randomization.sampleWithoutReplacement([1,2],1)[0];
//randomize version
if(version == 1){
     timeline.push(blocktrial_procedure_free, blocktrial_procedure_fixed);
}else {
     timeline.push(blocktrial_procedure_fixed, blocktrial_procedure_free);
};

Anyway, thank you :)

**Reply by becky-gilbert** [reply]
Great, glad you got it working!
